### DISCUSSION TITLE
ModularPipelines and Git Hooks

### DISCUSSION THREAD

**Discussion by Kevinf63** [thomhurst/ModularPipelines#475] [Ideas]
Hi @thomhurst et al, 

I haven't started to use your libraries just yet but I'm actually really interested in using them for a proof of concept I'm investigating around CI/CD, per their original definitions, with Git hooks (maybe Husky/Husky.net) and Azure Pipelines. Any experience using Git hooks or Husky?

The goal is ultimately:

- End users commit **directly to the trunk** (main branch)
- Our build pipeline is source controlled, modular, debuggable and can be locally run
- Common Git commands like commit and push will trigger the pipeline (Git Hooks)
- Git Push succeeding means the code is in a releaseable state and **validated by the pipeline**
- Catch code smells and vulnerabilities, perform linting, compilation, unit testing etc before Push
- Omit the need for an end user to **create branches**
- Omit the need for an end user to **open a pull request**
- Pull Request/Branches are most likely reserved for external bots like dependabot.

**Comment by thomhurst** [comment]
Heya @Kevinf63 !
My work has used Husky, and so I've got vague knowledge around git hooks. But essentially it's just a script defined within a file, that will be executed by git when you commit, right?

If there's no more to it than that, then modular pipelines should be able to meet your needs here. Your git hook just needs to be a `dotnet run MyPipeline.csproj` and that's it. The rest of your logic lies within your ModularPipelines project code, and you can just throw an exception within a module if something doesn't meet your criteria, and that should result in a non-zero exit code for the git hook and so it should fail and abort the commit.

All I would say, is that running dotnet requires compilation, so you will have some delay while it does this. On a build server you won't feel this as much, but locally on your machine, your commits might feel slow and laggy, so do consider that!
